### PR TITLE
map: Add a method to Unpin maps

### PR DESCRIPTION
**Pull request by aditighag**
This is an inverse of the Pin method. Also, add a method to get Map's pinned path.

Signed-off-by: Aditi Ghag aditi@cilium.io

### CODE REVIEW

**Review by lmb — COMMENTED** [review]
Thanks for your PR :) How would this work with a map being pinned in multiple places? Seems like calling Pin repeatedly will overwrite the previous pin state, so leave "stray" pinned maps. I think `Unpin(path string)` is the best we can do.

Regarding `quicktest`: is this something like `stretchr/testify`? This is my personal opinion, but I find that these libraries produce really inscrutable test output. Would you mind just using plain old `testing`?

**Comment by lmb** [comment]
The output from `quicktest` seems much more understandable than `stretchr/testify` at least:
```
=== RUN   TestFoo
    /home/lorenz/dev/quicktesttest/main_test.go:10:
        error:
          values are not equal
        got:
          bool(false)
        want:
          bool(true)
        stack:
          /home/lorenz/dev/quicktesttest/main_test.go:18
            qt.Assert(t, false, qt.Equals, true)
          /home/lorenz/dev/quicktesttest/main_test.go:10
            helper(t)
```

It also doesn't pull in a bazillion dependecies. Maybe it's not so bad? It's still an additional dependency, but oh well.

**Comment by aditighag** [comment]
>How would this work with a map being pinned in multiple places? Seems like calling Pin repeatedly will overwrite the previous pin state, so leave "stray" pinned maps.

This is an existing issue, no? Anyway I _think_ this should be easy to fix with the help of the new `IsPinned` method.

**Comment by aditighag** [comment]
> The output from `quicktest` seems much more understandable than `stretchr/testify` at least:
> 
> ```
> === RUN   TestFoo
>     /home/lorenz/dev/quicktesttest/main_test.go:10:
>         error:
>           values are not equal
>         got:
>           bool(false)
>         want:
>           bool(true)
>         stack:
>           /home/lorenz/dev/quicktesttest/main_test.go:18
>             qt.Assert(t, false, qt.Equals, true)
>           /home/lorenz/dev/quicktesttest/main_test.go:10
>             helper(t)
> ```
> 
> It also doesn't pull in a bazillion dependecies. Maybe it's not so bad? It's still an additional dependency, but oh well.

Yup, I like the pretty error reporting, and the fact that it exports only a handful of methods (less dependencies), quicktest trumps testify for our current use.

**Comment by aditighag** [comment]
>P.S.: What about Program.Unpin?

Haven't really thought about it. How is this related to `Map.Unpin` (mainly the focus of this PR), though?

**Comment by lmb** [comment]
> This is an existing issue, no?

```
m.Pin("/foo")
m.Pin("/bar")
m.Unpin()
```

With your PR, only /bar is unlinked. Basically, I don't share your assumption that a map is only ever pinned in one place. `IsPinned` doesn't fix that.

> How is this related to Map.Unpin (mainly the focus of this PR), though?

I'm playing the devils advocate. Program.Pin exists, and is used for the same purpose. Why should we add Map.Unpin but not Program.Unpin?

Maybe we're going at this in the wrong way: why do you need to unpin? From personal experience, a pinned map is never unpinned because it contains valuable state (that's why it's pinned). If you want to re-use a pinned map, you can use [MapOptions.PinPath](https://pkg.go.dev/github.com/cilium/ebpf#MapOptions) together with `LIBBPF_PIN_BY_NAME`. This has the effect that the library does the pinning and reusing of maps for you. If you then really want to get rid of your state that is  just `os.RemoveAll(PinPath)` away.

**Comment by aditighag** [comment]
> that a map is only ever pinned in one place.

I'm not aware of any use cases where a map is pinned to multiple paths. Can you elaborate on why we can't make this assumption?
/cc @joestringer 

>Why should we add Map.Unpin but not Program.Unpin?

By all means :) But that's outside the scope of this PR.

>If you then really want to get rid of your state that is just os.RemoveAll(PinPath) away.

It's not just that. With the current API design, we don't require users to pass the pinned path to `Unpin()`. There is a reason why the `Map`'s pinned path is embedded into the struct so that we can avoid error conditions where users pass a wrong path (e.g., another map's path) while calling `Unpin()` for a map. Thoughts?

**Comment by joestringer** [comment]
> I'm not aware of any use cases where a map is pinned to multiple paths. Can you elaborate on why we can't make this assumption?

The difficulty with library development is that the aim is to build something generic anyone could use, so it becomes important to the consider use cases outside the one we immediately have in front of us. If we have the foresight at this point to understand that there are two plausible use cases, then ideally we can come up with a sane way to handle both of those use cases.

I see @lmb's point here, currently there is no `Unpin()`, so assuming you have a use case, you can pin wherever you like and as many times as you like. It could be something as simple as "I have a program-friendly map naming scheme that can be generated from hashes/uuids, plus a user-friendly map naming scheme which makes it easier for users to interact with, and I want both maps to point to the same underlying object".  The "official" Unpin approach today would be, the user should simply remove the file from the filesystem through something like `os.Remove()`. Proper filesystem cleanup is completely up to the library user.

At a philosophical level though, if the library provides a `Pin()` function, then we are inherently saying that the library is responsible for handling filesystem association. That makes me wonder, why should `Unpin()` be different, just because under the hood it uses the `unlink()` syscall rather than `bpf()`? It's "nice" to have properly balanced "Add()"/"Remove()", "Create"/"Delete", or in this case "Pin()"/"Unpin()".

As a (slight) aside, actually looking at the `os.Remove()` [implementation](https://golang.org/src/os/file_unix.go?s=8292:8322#L265), if you pass a directory in instead of a path then it will attempt to delete the directory which seems like it could potentially have the wrong behaviour, hiding this behind a bpf library abstraction might make that kind of mistake more difficult to trigger (because we could ensure that the filepath is *not* a directory during the `Pin()`, and/or implement the `Unpin()` directly with `Unlink()` (though we probably don't want to get into os abstractions like that in this library, `os.Remove()` already does those bits for us). I think that's the kind of potholes that @aditighag suggests we could help library users to avoid.

Looking at the code proposal, one natural way to handle multiple mappings would be:

```
m1 := NewMap(...)
// Optional; could be part of Spec or could be manually done
m2.Pin("/first/path")
m2 := m.Clone()
m2.Pin("/second/path")
// ... use the maps
m.Unpin()
// ... keep m2 around for a while if you want
m2.Unpin()
```

Then, if you really care about separately handling the paths because your use case demands that the paths have different lifecycles, you keep two structures around and interact with them independently. To make this more clean, one might argue to take it one step further and force the `Pin()` to tie to a single path, ie if you `Pin()` an already-pinned map, it either fails with "ErrorAlreadyPinned" or moves the pinning to the new location. Then it would be difficult for the library user to shoot themselves in the foot by leaving old pinnings on the filesystem. Downside is a little extra memory usage, but for the rich representation of the map, maybe that's not so terrible?


> Maybe we're going at this in the wrong way: why do you need to unpin? From personal experience, a pinned map is never unpinned because it contains valuable state (that's why it's pinned). If you want to re-use a pinned map, you can use MapOptions.PinPath together with LIBBPF_PIN_BY_NAME. This has the effect that the library does the pinning and reusing of maps for you. If you then really want to get rid of your state that is just os.RemoveAll(PinPath) away.

The Cilium use case is primarily around upgrade or reconfiguration. One version of Cilium defines the map in one way (with certain attributes like size for instance), then the new version defines the map in a different way. We want to reuse the path but redefine the attributes. I think the following workflow is fairly reasonable (though I didn't spend too long looking into it):

```
fd := os.Open(...)
m := NewMapFromFD(fd)
// Check the attributes match what we want them to be, if they're good we use `m`, if not we want to delete/recreate it (and potentially run some special logic to safely migrate map contents).
m.Unpin()
```

A slight variation on this case is that we make sure there is a backup of the currently-active map, then we create the new map and populate it, then we switch the BPF program over to use the new map, then once everything is successful, we unpin the old one. If some of those steps were unsuccessful (eg prog load fails), then we can roll back all the steps. `cilium/cilium/pkg/bpf/map_linux.go:OpenParallel()` is one example, there's also `pkg/datapath/ipcache/listener.go:garbageCollect()` with the `shuffleMaps()` for another example. Some of that code could improve from more reuse of this library, but either way they'll need to actually handle the `Unpin()` of the old map. These are not required to be during upgrade either, the `garbageCollect()` case is recreating the same map in-place to work around a kernel behaviour on older kernels.

**Comment by aditighag** [comment]
Thanks for the detailed answer, Joe. 

Sure, it makes sense to not make the assumption with respect to how many times a map can be pinned. 

> 
> Looking at the code proposal, one natural way to handle multiple mappings would be:
> 
> ```
> m1 := NewMap(...)
> // Optional; could be part of Spec or could be manually done
> m2.Pin("/first/path")
> m2 := m.Clone()
> m2.Pin("/second/path")
> // ... use the maps
> m.Unpin()
> // ... keep m2 around for a while if you want
> m2.Unpin()
> ```
> 
> Then, if you really care about separately handling the paths because your use case demands that the paths have different lifecycles, you keep two structures around and interact with them independently. To make this more clean, one might argue to take it one step further and force the `Pin()` to tie to a single path, ie if you `Pin()` an already-pinned map, it either fails with "ErrorAlreadyPinned" or moves the pinning to the new location. Then it would be difficult for the library user to shoot themselves in the foot by leaving old pinnings on the filesystem. Downside is a little extra memory usage, but for the rich representation of the map, maybe that's not so terrible?

Yes, the current design of the `Clone` API can be used to support multiple pinned paths. Either we could document that users can use `Clone` to pin a map to multiple paths, or we can facilitate that behind the scenes. If it's the latter, I'm inclined to keep the current `Pin` method as-is (aligned with the most common use case), and introduce a new method `PinWithClone` to make the intent more clear.
@joestringer @lmb Do you prefer one to the other?

File: map.go [hunk 1]
```diff
@@ -688,7 +697,41 @@ func (m *Map) Clone() (*Map, error) {
...
+	}
+	err := os.Remove(m.pinState.path)
+	if err == nil {
+		m.pinState = mapPinnedState{}
+	}
+	return err
+}
+
+// IsPinned returns (true, pinned path) if the map is pinned, (false, "") otherwise.
+func (m *Map) IsPinned() (bool, string) {
```
[the remark this answers is not in the document]
Comment (aditighag) [reply map.go:730]: As the comment above suggests, it's an invariant - If `pinned == true`, then `path = ""` (default value of `String`). The intention is to not to keep users guessing whether the map is pinned or not when `path == ""`.

**Comment by joestringer** [comment]
I don't think we need extra functions like `PinWithClone()`. If there's a simple, straightforward way to achieve this for users who care by doing `m2 := m.Clone()` + `m2.Pin()` then that would satisfy the use case.

**Comment by aditighag** [comment]
> I don't think we need extra functions like `PinWithClone()`. If there's a simple, straightforward way to achieve this for users who care by doing `m2 := m.Clone()` + `m2.Pin()` then that would satisfy the use case.

Yup, that was my preference too. I've documented the behavior. PTAL.

**Review by lmb — CHANGES_REQUESTED** [review]
Thanks for the detailed discussion @joestringer and thanks for following up @aditighag. You've convinced me that `Pin()` and `Clone()` is acceptable if someone really wants to pin a map multiple times.

@aditighag the PR currently doesn't handle `LoadPinnedMap`, I think any map returned from the function should have it's `Unpin()` method work.

I also played around with the API a bit, I think there are small things we should address:
```
m.Pin("") // not ok, probably bogus (1)

m.Pin(/foo)
m.Pin(/foo) // ok, pin is idempotent

m.Pin(/foo)
m.Pin(/bar) // not ok, already pinned

m.Pin(/foo)
m.Unpin()
m.Pin(/bar) // ok, not pinned

m.Pin(/foo)
m.Unpin()
m.Unpin() // ok, unpin is idempotent

m = LoadPinnedMap(/foo)
m.Pin() // not ok, already pinned

m = LoadPinnedMap(/foo)
m.Unpin() // ok

m1 = LoadPinnedMap(/foo)
m2 = LoadPinnedMap(/foo)
m1.Unpin()
m2.Unpin() // not ok, file was removed by m1 (2)

m1.Pin()
m2 = m1.Clone()
m1.Unpin()
m2.Unpin() // ok, not pinned (3)
```

Issue (1) goes away by just having `pinPath` as the state, treating the empty path as meaning "not pinned" and forbidding `Pin("")`.

Issue (2) is a bit more subtle. Since both m1 and m2 come from `LoadPinnedMap` they can be unpinned. However, calling Unpin on both will return an error in one of them. I think this is compounded by issue (3), which is that Unpinning a cloned map on the other hand doesn't return an error. Since `Unpin` so far is idempotent, maybe it makes sense to treat `ENOENT` from `os.Remove` as no error? Then the behaviour of (2) and (3) is the same.

Finally, why don't you want to add the same functions to `Program`? I want to keep the API between the two consistent, and it's not a lot of work.

@joestringer thanks for the pointers to the cilium code base. Some interesting points:

https://github.com/cilium/cilium/blob/1ef686bfb92bb3cc7ee1ac5a6d5d706835e34806/pkg/datapath/ipcache/listener.go#L232-L247

Let's imagine this uses the new API:
```Go
func shuffleMaps(a, b Map) error {
    os.Rename(a.Path(), "backup")
    os.Rename(b.Path(), a.Path())
    // not sure what happens to "backup"?
}
```

Since we've tied a path to a map, both a and b are inconsistent. `a.Unpin` will remove b, and `b.Unpin` produces an error. I don't see how to make this safe.

https://github.com/cilium/cilium/blob/c158fb169742923d7e97b52fde98388198be8ada/pkg/bpf/map_linux.go#L420-L444

If I understand correctly this is equivalent to:
```
func OpenParallel(a, b Map) {
    path := a.Path()
    a.Unpin()
    b.Pin(path)
}
```

This works fine, I think. Note that skipping `a.Unpin` because we always want there to be a pinned map is not safe, because then `a.Unpin` would remove b. In general, playing games with `.Path()` is a hazard, so I wonder if there is a way to get away without it?

I also found https://github.com/cilium/cilium/blob/c158fb169742923d7e97b52fde98388198be8ada/pkg/bpf/map_linux.go#L498-L502 which makes me curious: why do you pin the map if its not persistent?

File: map.go [hunk 2]
```diff
@@ -680,15 +689,53 @@ func (m *Map) Clone() (*Map, error) {
...
+		m.pinState.path = fileName
+	}
+	return err
+}
+
+// Unpin removes the persisted state for the map.
+//
+// Unpinning a nil or un-pinned Map returns nil.
+func (m *Map) Unpin() error {
+	if m == nil || !m.pinState.pinned {
```
[the remark this answers is not in the document]
Comment (aditighag) [reply map.go:723]: Well, more importantly, if `Unpin` is called, then `m` can't be nil. I'll fix this.

**Review by aditighag — COMMENTED** [review]
Thanks for highlighting `LoadPinnedMap`, @lmb, good point. I'll think about how to tackle (2).

**Comment by joestringer** [comment]
> Let's imagine this uses the new API:

> func shuffleMaps(a, b Map) error {
    os.Rename(a.Path(), "backup")
    os.Rename(b.Path(), a.Path())
    // not sure what happens to "backup"?
}

> Since we've tied a path to a map, both a and b are inconsistent. a.Unpin will remove b, and b.Unpin produces an error. I don't see how to make this safe.

Right, @aditighag this is an interesting point on the design. By adding `Unpin()` and holding path state, we *force* library users to interact with the library to change the pinning on the filesystem. If library users otherwise interact with the underlying filesystem like the above, then the `Unpin()` will fail.

To implement 1:1, the above would need to become something like:

```
path := oldMap.Path()
oldMap.Unpin()
newMap.Pin(path)
```

(plus the rollback logic to re-pin `oldMap` upon failure).


I think that the idea would be to replace the whole `shuffleMaps` stuff instead with two in-memory `*ebpf.Map`, each of which has an open fd, and you basically unpin

**Comment by joestringer** [comment]
> Let's imagine this uses the new API:

> func shuffleMaps(a, b Map) error {
    os.Rename(a.Path(), "backup")
    os.Rename(b.Path(), a.Path())
    // not sure what happens to "backup"?
}

> Since we've tied a path to a map, both a and b are inconsistent. a.Unpin will remove b, and b.Unpin produces an error. I don't see how to make this safe.

Right, @aditighag this is an interesting point on the design. By adding `Unpin()` and holding path state, we *force* library users to interact with the library to change the pinning on the filesystem. If library users otherwise interact with the underlying filesystem like the above, then the `Unpin()` will fail.

To implement 1:1, the above would need to become something like:

```
path := oldMap.Path()
// Depending on whether we define that `Pin()` has move semantics or create semantics (ie fail if it is already pinned), unpin the old path
oldMap.Unpin()
oldMap.Pin("backup")
newMap.Pin(path)
```

Strictly speaking for the actual use case above, we could probably get away with just an `oldMap.Unpin()` + `newMap.Pin()`, I think it's just trying to ensure that there are traces on the filesystem of the ongoing operation, in case anything goes wrong (kill the daemon in between operations, perhaps?)

**Comment by aditighag** [comment]
I've pushed a new update to accommodate the test cases involving `LoadPinnedMap`. The latest changes ensure that users won't get inconsistent view of a map's pinned state as long as they have used the internal APIs of the library. 

> Let's imagine this uses the new API:
> 
> func shuffleMaps(a, b Map) error {
>     os.Rename(a.Path(), "backup")
>     os.Rename(b.Path(), a.Path())
>     // not sure what happens to "backup"?
> }
With the new APIs, the function calls will translate to realized.Unpin(), realized.Pin("backup"), pending.Unpin(), pending.Pin("realized").  

>  If library users otherwise interact with the underlying filesystem like the above, then the Unpin() will fail.

I've changed the behavior such that it won't.

**Comment by aditighag** [comment]
Re: `Unpin` for `prog`, I went over prog.go, looks like the logic is likely to be same as `Map.Unpin`. Once we finalize the API design, I'll implement a common construct for both `Map` and `Prog`.

**Comment by ti-mo** [comment]
> Since Unpin so far is idempotent, maybe it makes sense to treat ENOENT from os.Remove as no error? Then the behaviour of (2) and (3) is the same.

Agreed, if ENOENT is received, that means `Unpin()`'s goal is already accomplished, although as a user I might want to check the status of the unpin operation. Not sure if this would come in handy.

> Note that skipping a.Unpin because we always want there to be a pinned map is not safe, because then a.Unpin would remove b. In general, playing games with .Path() is a hazard, so I wonder if there is a way to get away without it?

IMO the safer way would be calling `os.Rename()` under the hood and keeping the map's internal pin path state in sync. Would be cleaner for the caller, too. Instead of juggling `Unpin()` and `Pin()`s to recreate a map, simply:

```
old.Pin("backup")
new.Pin(old_path)
```

Error handling becomes less involved for the caller because there are less moving parts. If `old.Pin("backup")` succeeds, there's a guarantee that the fs was updated and that the new path was committed to `old`'s struct.

Re: `prog.Unpin()`, why not tackle that in a separate PR?

**Review by lmb — CHANGES_REQUESTED** [review]
@ti-mo you're saying that `m.Pin("/a"); m.Pin("/b")` should be valid, and that the latter call should do an `os.Rename` instead of re-pinning? Interesting idea, together with IsPinned this would probably give good flexibility.

It also gets rid of the following hazard (from @aditighag example):

```
realized.Unpin()
// crash here loses realized if there is no other reference
realized.Pin("backup")
pending.Unpin()
// same for pending
pending.Pin("realized")
```

However, we still have the following problem:

```
old.Pin("backup")
// crash here makes old not discoverable so a restart will fail
new.Pin(old_path) // NB: old_path is not accessible from old anymore
```

I'm a bit fuzzy what the backup is used for (since it's not removed?) so I don't know if this is an issue.

**Comment by ti-mo** [comment]
```
old.Pin("backup")
// crash here makes old not discoverable so a restart will fail
```
Assuming `old.Pin("backup")` was successful, `backup` maintains a ref to the map, so the map will outlive the process in case of a crash. If the process comes back up, it can/should discover an existing map by opening `backup` first, or it can clean up and start from scratch depending on the use case.

`Pin()` using an atomic fs operation removes all risk of the process terminating while it's holding the only map ref, so there shouldn't be any data loss.

**Comment by lmb** [comment]
@aditighag ping, are you still interested in this?

**Comment by aditighag** [comment]
@lmb Yes, I'm still working on it. I'll follow-up shortly.

**Comment by aditighag** [comment]
Pushed a new commit with the following main changes - 

- `Pin` method allows overriding previously pinned path. This was done to accommodate the Cilium like use cases where map paths are updated for backup purposes. 
- `IsPinned` now only checks if a map has a non-empty path or not.

Thanks!

**Review by lmb — APPROVED** [review]
LGTM!

**Comment by aditighag** [comment]
@lmb Thanks for your detailed review. Can we get this merged? I'll send another PR with prog pinning support.

**Pull request merged**